collect_app/src/main/java/org/odk/collect/android/external/FormUriActivity.kt

```diff
@@ -317,6 +328,21 @@ private class FormUriViewModel(
         }
     }
 
+    private fun assertFormsUpdateNotInProgress(): String? {
+        val projectId = projectsDataService.getCurrentProject().uuid
+        val formsLock = changeLockProvider.create(projectId).formsLock
+        return if (formsLock.isLocked()) {
```

Comment:
I'm realising the "check and then lock" needs to atomic here. Otherwise, the lock could be acquired by another thread in between the `isLocked` and `lock` calls.

I think instead of `isLocked`/`lock` methods, we should add a `tryLock(): Boolean` to `ChangeLock`. This would just wrap `ReentrantLock#tryLock` in `ReentrantLockChangeLock` and would allow the code to look like this:

```kotlin
return if (formsLock.tryLock()) {
    null
} else {
    resources.getString(string.cannot_open_form_because_of_forms_update)
}
```

Reply:
Done.